src/idl_parser.cpp

```diff
@@ -86,26 +86,29 @@ CheckedError Parser::Error(const std::string &msg) {
 
 inline CheckedError NoError() { return CheckedError(false); }
 
+inline std::string OutOfRangeErrorMsg(int64_t val, const std::string& op,
+                                      int64_t limit) {
+  const std::string cause = NumToString(val) + op + NumToString(limit);
+  return "constant does not fit (" + cause + ")";
+}
+
 // Ensure that integer values we parse fit inside the declared integer type.
-CheckedError Parser::CheckBitsFit(int64_t val, size_t bits) {
-  // Left-shifting a 64-bit value by 64 bits or more is undefined
-  // behavior (C99 6.5.7), so check *before* we shift.
-  if (bits < 64) {
-    // Bits we allow to be used.
-    auto mask = static_cast<int64_t>((1ull << bits) - 1);
-    if ((val & ~mask) != 0 &&  // Positive or unsigned.
-        (val |  mask) != -1)   // Negative.
-      return Error("constant does not fit in a " + NumToString(bits) +
-                   "-bit field");
-  }
-  return NoError();
+CheckedError Parser::CheckInRange(int64_t val, int64_t min, int64_t max) {
+  if (val < min)
```

Comment:
I did have that when I first worked on it, but I didn't like the repeated `val < min` in order to have separate error messages.
See https://github.com/jasonstubbs/flatbuffers/blob/0789a9c2bb39a33bcf1b3259825391956ffeaa04/src/idl_parser.cpp#L91-L97

Changing the error message to work in both cases would be fine to.
See https://github.com/jasonstubbs/flatbuffers/blob/fix/4227-alternate/src/idl_parser.cpp#L91-L93

I'd really prefer to show the type (eg, ulong) rather than the numeric limit that is being exceeded, but I can't see a way to do that without making it even more convoluted.

I'm happy to change it to whichever way you prefer.